modules/detection/TrackVADEmitter.js

```diff
@@ -0,0 +1,231 @@
+import RTC from '../RTC/RTC';
+import { VAD_SCORE_PUBLISHED } from './DetectionEvents';
+
+/**
+ * Connects an audio JitsiLocalTrack to a vadProcessor using WebAudio ScriptProcessorNode.
+ * Once an object is created audio from the local track flows through the ScriptProcessorNode as raw PCM.
+ * The PCM is processed by the injected vad module and a voice activity detection score is obtained, the
+ * score is published to consumers via an EventEmitter.
+ * After work is done with this service the destroy method needs to be called for a proper cleanup.
+ *
+ * @fires VAD_SCORE_PUBLISHED
+ */
+export default class TrackVADEmitter extends EventEmitter {
+    /**
+     * Constructor.
+     *
+     * @param {number} procNodeSampleRate - Sample rate of the ScriptProcessorNode. Possible values  256, 512, 1024,
+     *  2048, 4096, 8192, 16384. Passing other values will default to closes neighbor.
+     * @param {Object} vadProcessor - VAD processor that allows us to calculate VAD score for PCM samples.
+     * @param {JitsiLocalTrack} jitsiLocalTrack - JitsiLocalTrack corresponding to micDeviceId.
+     */
+    constructor(procNodeSampleRate, vadProcessor, jitsiLocalTrack) {
+        super();
+
+        /**
+         * Sample rate of the ScriptProcessorNode.
+         */
+        this._procNodeSampleRate = procNodeSampleRate;
+
+        /**
+         * VAD Processor that allows us to calculate VAD score for PCM samples
+         */
+        this._vadProcessor = vadProcessor;
+
+        /**
+         * The JitsiLocalTrack instance.
+         */
+        this._localTrack = jitsiLocalTrack;
+
+        /**
+         * Buffer to hold residue PCM resulting after a ScriptProcessorNode callback
+         */
+        this._bufferResidue = new Float32Array([]);
+
+        /**
+         * The AudioContext instance with the preferred sample frequency.
+         */
+        this._audioContext = new AudioContext({ sampleRate: vadProcessor.getRequiredPCMFrequency() });
+
+        /**
+         * PCM Sample size expected by the VAD Processor instance. We cache it here as this value is used extensively,
+         * saves a couple of function calls.
+         */
+        this._vadSampleSize = vadProcessor.getSampleLength();
+
+        /**
+         * Event listener function that will be called by the ScriptProcessNode with raw PCM data, depending on the set
+         * sample rate.
+         */
+        this._onAudioProcess = this._onAudioProcess.bind(this);
+
+        this._initializeAudioContext();
+    }
+
+    /**
+     * Factory method that sets up all the necessary components for the creation of the TrackVADEmitter.
+     *
+     * @param {string} micDeviceId - Target microphone device id.
+     * @param {number} procNodeSampleRate - Sample rate of the proc node.
+     * @returns {Promise<TrackVADEmitter>} - Promise resolving in a new instance of TrackVADEmitter.
+     */
+    static create(micDeviceId, procNodeSampleRate, vadProcessor) {
```

Comment:
So I actually thought about this a bit, because the class component depends on a async loaded resource we need would need to use a promise inside the c'tor which gave me a lot of headaches when it came to managing state, so did some reading and came across this https://stackoverflow.com/questions/24398699/is-it-bad-practice-to-have-a-constructor-function-return-a-promise and apparently using a promise inside a c'tor is a no go.